api/scanner/media_encoding/executable_worker/executable_worker_test.go

```diff
@@ -0,0 +1,91 @@
+package executable_worker_test
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/photoview/photoview/api/scanner/media_encoding/executable_worker"
+	"github.com/photoview/photoview/api/test_utils"
+)
+
+func TestMain(m *testing.M) {
+	os.Exit(test_utils.IntegrationTestRun(m))
+}
+
+func setPathWithCurrent(paths ...string) func() {
+	_, file, _, ok := runtime.Caller(0)
+	if !ok {
+		return func() {}
+	}
+
+	base := filepath.Dir(file)
+
+	for i, path := range paths {
+		paths[i] = filepath.Join(base, path)
+	}
+
+	originalPath := os.Getenv("PATH")
+	os.Setenv("PATH", strings.Join(paths, ":"))
+
+	return func() {
+		os.Setenv("PATH", originalPath)
+	}
+}
+
+func TestMagickWorkerNotExist(t *testing.T) {
+	done := setPathWithCurrent()
+	defer done()
+
+	executable_worker.InitializeExecutableWorkers()
+
+	if executable_worker.MagickCli.IsInstalled() {
+		t.Error("MagickCli should not be installed, but is found:", executable_worker.MagickCli)
+	}
+}
+
+func TestMagickWorkerIgnore(t *testing.T) {
+	done := setPathWithCurrent("./testdata/bin")
+	defer done()
+
+	org := os.Getenv("PHOTOVIEW_DISABLE_RAW_PROCESSING")
+	os.Setenv("PHOTOVIEW_DISABLE_RAW_PROCESSING", "true")
+	defer os.Setenv("PHOTOVIEW_DISABLE_RAW_PROCESSING", org)
+
+	executable_worker.InitializeExecutableWorkers()
+
+	if executable_worker.MagickCli.IsInstalled() {
+		t.Error("MagickCli should not be installed, but is found:", executable_worker.MagickCli)
+	}
+}
+
+func TestMagickWorker(t *testing.T) {
+	done := setPathWithCurrent("./testdata/bin")
+	defer done()
+
+	executable_worker.InitializeExecutableWorkers()
+
+	if !executable_worker.MagickCli.IsInstalled() {
+		t.Error("MagickCli should be installed")
+	}
+
+	t.Run("Failed", func(t *testing.T) {
+		err := executable_worker.MagickCli.EncodeJpeg("input", "output", 0)
+		if err == nil {
+			t.Fatalf("MagickCli.EncodeJpeg(\"input\", \"output\", 0) = nil, should be an error.")
+		}
+		if got, want := err.Error(), "^encoding image with \".*?/testdata/bin/convert .*?\" error: .*$"; !regexp.MustCompile(want).MatchString(got) {
```

Comment:
I'm lost in this code)))
I see that the `.*?/testdata/bin/convert` is the stub replacement, which shouldn't do anything, but cannot understand what is tested and how is it used here))

Most probably using asserts would make it much more clear...

Reply:
I don't want to use any assert library. They always hide some details and are hard to understand after a while.
Here, I am checking the error message with the command line. Since the path of the repo is not fixed, the only fixed path is `/testdata/bin/convert`.

Reply:
Hmm, do you see something hidden and not clear in my tests (referenced in the previous comment)?
From my experience, using the correct assert types with enough explanation (failure message) makes tests much more clear.
Please give an example of hidden details to let me understand your point

Reply:
For example, https://github.com/kkovaletp/photoview/blob/hw-gpu-support/api/scanner/media_encoding/executable_worker/darktable_worker_test.go#L22 this should be `require.True()`, because if it fails, all following checks are non-sense and it should fail the whole test immediately. All assert libraries change those words in `testing.T` and it's hard to know the details without checking the code.

The second problem is that libraries fix the error message. It's hard to customize that message. If I compare two multi-line strings, I hope to print a diff, but not just blob content of two strings. There are more cases but I can't remember them all. I believe the failure message is important for maintenance.

Third is the order of parameters. `assert.Equal(got, want)` or `assert.Equal(want, got)`?

The last, it reduces dependencies.

Following with [Go official wiki](https://go.dev/wiki/TestComments#assert-libraries) and don't use any assert libraries.

Reply:
Wow(
Now I see your point.
Before this project, I wrote tests in Java and Python. There are no such severe issues and assert libraries are mature and stable.

While the 1st case is just an issue, caused by the incorrect type of assert used (and can be easily fixed on code review), it is also (in this particular case) not a big deal if other asserts will run even if this one fails. However, it shows the level of complexity to choose the right assert from the long list. I'd better have a single assert with optional behavior modification parameters.

The 2nd case describes the issue, I noticed in the test execution results earlier as well. Java and Python assert libs together with testing framework provide much better results reporting: results provide more info to understand the root cause and debug the code, better structured visually (but this is not related to the assert lib).

The 3rd one is tricky, but IDE suggests the order by showing the function parameters in the tooltip.

In general, assert libs provide an abstraction layer to developers, allowing them to write less code, clear code, and concentrate on test logic instead of the language-specific implementation. But if GO doesn't provide good assert libs, maybe it makes sense to not use them at all... However, really good unit tests, which cover the code with good quality of testing and coverage level, would be too complex and too big in this case...

Reply:
I believe a rich-feature assert library began in Ruby and other languages copy the idea. I found these problems also exist in these languages, but the community tolerates them.

I don't insist on it and want to throw this discussion out. I see `photoview` already uses https://github.com/stretchr/testify. If you insist on it, I can update the code.

Reply:
No, I'm not insisting) I'm fine as long as we have good enough coverage and the unit tests contain correct checks, verifying the right things.
I'd rely on your expert recommendations on how it is better to implement those checks. I just wanted to discuss some negative aspects and uncover the increasing complexity of the testing code - just to make sure that this is done intentionally and that all pros and cons are taken into account.

In this case, if I'll contribute some changes to the code, I'll try to follow this tests' architecture, but I'm not sure whether I'll be able to maintain the same level of code quality. I'll try and let's see the result)

Reply:
SG. I resolved this conversation now.